integration-test/http/http_system_test.c

```diff
@@ -0,0 +1,362 @@
+    /* Represents a response returned from an HTTP server. */
+    HTTPResponse_t response;
+    /* Represents header data that will be sent in an HTTP request. */
+    HTTPRequestHeaders_t requestHeaders;
+
+    assert( pMethod != NULL );
+    assert( pPath != NULL );
+
+    /* Initialize all HTTP Client library API structs to 0. */
+    ( void ) memset( &requestInfo, 0, sizeof( requestInfo ) );
+    ( void ) memset( &response, 0, sizeof( response ) );
+    ( void ) memset( &requestHeaders, 0, sizeof( requestHeaders ) );
+
+    /* Initialize the request object. */
+    requestInfo.pHost = SERVER_HOST;
+    requestInfo.hostLen = SERVER_HOST_LENGTH;
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = strlen( pMethod );
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = strlen( pPath );
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. */
+    requestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Set the buffer used for storing request headers. */
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    /* Initialize the response object. The same buffer used for storing
+     * request headers is reused here. */
+    response.pBuffer = userBuffer;
+    response.bufferLen = USER_BUFFER_LENGTH;
+
+    LogDebug( ( "Sending HTTP %.*s request to %.*s%.*s...",
+                ( int32_t ) requestInfo.methodLen, requestInfo.method,
+                ( int32_t ) SERVER_HOST_LENGTH, SERVER_HOST,
+                ( int32_t ) requestInfo.pathLen, requestInfo.pPath ) );
+
+    /* Send request to HTTP server. If a network error is found, retry request for a
+     * count of MAX_RETRY_COUNT. */
+    do
+    {
+        /* Since the request and response headers share a buffer, request headers should
+         * be re-initialized after a failed request. */
+        httpStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                          &requestInfo );
+        TEST_ASSERT_EQUAL( HTTP_SUCCESS, httpStatus );
+        TEST_ASSERT_NOT_EQUAL( 0, requestHeaders.headersLen );
+
+        LogDebug( ( "Request Headers:\n%.*s\n"
+                    "Request Body:\n%.*s\n",
+                    ( int32_t ) requestHeaders.headersLen,
+                    ( char * ) requestHeaders.pBuffer,
+                    ( int32_t ) REQUEST_BODY_LENGTH, REQUEST_BODY ) );
+
+        httpStatus = HTTPClient_Send( pTransportInterface,
+                                      &requestHeaders,
+                                      ( uint8_t * ) REQUEST_BODY,
+                                      REQUEST_BODY_LENGTH,
+                                      &response,
+                                      0 );
+
+        if( httpStatus == HTTP_NETWORK_ERROR )
+        {
+            LogDebug( ( "A network error has occured, retrying request." ) );
+            resetTest();
```

Comment:
Is this a Unity API? If so, has this been tested?

Reply:
`resetTest()` gets generated with the test runner -- the function itself just calls `tearDown()` and `setUp()`, which in this case are responsible for disconnecting and reconnecting the network. I have only tested it locally with some false network errors, but it had been working as expected, and the CMock docs clarify that it can be called from within tests.